util: Abort on failing CHECK_NONFATAL in debug builds

maflcko opened the pull request:
A failing `CHECK_NONFATAL` will throw an exception. This is fine and even desired in production builds, because the program may catch the exception and give the user a way to easily report the bug upstream.

However, in debug development builds, exceptions for internal bugs are problematic:

* The exception could accidentally be caught and silently ignored
* The exception does not include a full stacktrace, possibly making debugging harder

Fix all issues by turning the exception into an abort in debug builds.

This can be tested by reverting the hunks to `src/rpc/node.cpp` and `test/functional/rpc_misc.py` and then running the functional or fuzz tests.

Commit:
refactor: Set G_ABORT_ON_FAILED_ASSUME when G_FUZZING_BUILD

This does not change behavior, but documents that
G_ABORT_ON_FAILED_ASSUME is set when G_FUZZING_BUILD.

Commit:
util: Abort on failing CHECK_NONFATAL in debug builds

This requires adjusting some tests to force exceptions over aborts, or
accept either exceptions or aborts.

Also, remove a fuzz test in integer.cpp that is mostly redundant with
the unit test added in the prior commit.

ryanofsky commented:
Concept ACK. Nice idea, and it does seem useful to have a macro checking for unexpected but not very serious conditions by throwing an exception that gets reported in release builds but is a fatal error in debug builds. And current uses of the macro seem like good candidates for that behavior.

The only possible issues I see are:

(1) The name CHECK_NONFATAL doesn't make a lot of sense anymore, now triggering fatal errors when it literally says "nonfatal" in the name.
(2) It is now more cumbersome to write unit tests checking for these conditions since they require a release build to run.

Both could be addressed in followups. Issue (2) could be addressed by having a `g_abort_hook` or similar hook allowing specific unit tests to write custom code to check for these errors if they want. (This could also be used to replace the `g_debug_lockorder_abort` variable which does something similar.)

IMO, issue (1) would be nice to address by coming up with a better designed set of checking macros and starting to use them. I think it could be good to have a:

- `CHECK` to check conditions and abort if false in all builds
- `DCHECK` to do the same but be compiled out of release builds,
- `CHECK_LOG` to log an "internal bug detected please report" type log message in release builds, and abort in debug builds
- `CHECK_THROW` to throw an exception in release builds, and abort in log builds.

Then, current `assert` uses could become `CHECK`, current `Assume` uses in hotspots could become `DCHECK`, majority of other `Assume` uses could become `CHECK_LOG`, and current `CHECK_NONFATAL` uses could become `CHECK_THROW`.

Just a thought though. Maybe current names are not a real problem, and naming shouldn't block this PR in any case.

Commit:
test: Allow testing of check failures

This allows specific tests to mock the check behavior to consistently
use exceptions instead of aborts for intentionally failing checks in all
build configurations.

achow101 commented:
Concept ACK

ryanofsky reviewed (approved):
Code review ACK faae9a294798c6808ec82f827385d920f8d697cf. I think this is a good change. It makes sense conceptually to have check macros that always abort in debug builds, but do different things depending on cost of the check & severity of the error in release builds.

re: https://github.com/bitcoin/bitcoin/pull/32588#issuecomment-2903414680

> I don't think this solves issue (1). Instead of NONFATAL being inaccurately named in debug builds, it will now be THROW, because the check is neither nonfatal nor throwing in debug builds.

IMO it does solve it, because the current issue is that the macro is literally doing the thing its name says it will not do (trigger a fatal error). By contrast, I don't think it is a problem for function name to just describe its primary purpose and not everything else it may do. No need to solve everything here though. Current PR seems like a step forward.

ryanofsky commented on src/rpc/node.cpp:
```diff
@@ -295,10 +294,6 @@ static RPCHelpMan echo(const std::string& name)
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
 {
-    if (request.params[9].isStr()) {
-        CHECK_NONFATAL(request.params[9].get_str() != "trigger_internal_bug");
```
re: https://github.com/bitcoin/bitcoin/pull/32588#discussion_r2223179634

> restored this, and extended the workaround in the fuzz test, and wrote unit tests.

Changes look nice, but to be sure there still isn't a test that replaces the python test, explicitly passing a "trigger_internal_bug" argument to the echo RPC and ensuring that it returns an "Internal bug detected" message in response. The fuzz test handles this case, but doesn't try to trigger it, so won't fail if the CHECK_NONFATAL` line is removed from the echo RPC.

Could declare a test_only_CheckFailuresAreExceptionsNotAborts instance inside the echo RPC to keep the python test intact (though would need need to make the global atomic). Or could add a simple unit test replacing the python test. Dropping the coverage also seems ok if you prefer.

ryanofsky reviewed (approved):
Code review ACK fa68bda5f82887f95cc0313385c8e3a97eaa9967. Looks good! Since last review added some more test coverage and reverted some unneeded changes.

ryanofsky commented on src/test/fuzz/integer.cpp:
```diff
@@ -255,9 +255,4 @@ FUZZ_TARGET(integer, .init = initialize_integer)
         }
     }
-
-    try {
-        CHECK_NONFATAL(b);
-    } catch (const NonFatalCheckError&) {
-    }
```
In commit "test: Allow testing of check failures" (fa8251bb883300bb84cad5486468d07b5b1b7322)

Not important, but I think it makes more sense to drop this test in the next commit instead dropping it here, because the next commit is what actually breaks this test and the claim that this fuzz test is "redundant to the new unit test" is not 100% accurate, since the fuzz test and unit test execute with different `CheckFailuresAreExceptions` values.

ryanofsky replied:
> I think it is 100% accurate, because `test_only_CheckFailuresAreExceptionsNotAborts` does not affect `CHECK_NONFATAL` in commit [fa8251b](https://github.com/bitcoin/bitcoin/commit/fa8251bb883300bb84cad5486468d07b5b1b7322), so the fuzz test and unit test are 100% identical and redundant in this commit.

I wouldn't say it's 100% accurate to say that two tests have identical coverage just because the they have the same results in one commit, especially when the results immediately diverge in the next commit, where one of the tests starts to fail and the other succeeds. But if you think this statement is 100% accurate, I think you can make it 101% accurate if you mention the change in the next commit, or just remove the test in the commit which breaks it. Also fine with the current approach, but wanted to clarify!

ryanofsky reviewed (approved):
Code review ACK fac50ee1f9e577527650397891d356bca6316321, just rebased, added back functional test, and tweaked fuzz test since last review.

Overall this looks good and conceptually I like this change because it makes all the checking macros do the exact same thing in debug builds and abort, only having varying behavior in release builds.

ryanofsky reviewed (approved):
Code review ACK fa37153288ca420420636046ef6b8c4ba7e5a478, just catching subprocess.CalledProcessError in test fixing up a comment since last review

stickies-v reviewed (approved):
ACK fa37153288ca420420636046ef6b8c4ba7e5a478
